File: src/mev/sensors/searcher_density_sensor.py
"""
Searcher Density Sensor - MEV Bot Activity Detection

Extracted from AxionCitadel - Operation First Light validated
Monitors MEV bot activity to adjust risk calculations dynamically

Metrics:
1. MEV transaction ratio (MEV txs / total txs)
2. Sandwich attack likelihood (gas price variance analysis)
3. Bot clustering (high-gas-paying address concentration)

Returns: Searcher density score (0.0 - 1.0)
"""

from web3 import Web3
from typing import List, Dict, Set, Tuple
import statistics


class SearcherDensitySensor:
    """Real-time MEV bot activity monitoring for risk adjustment"""
    
    def __init__(self, rpc_url: str, routers: List[str], window_size: int = 10, weights: Tuple[float, float, float] = (0.4, 0.4, 0.2)):
        """
        Initialize searcher density sensor
        
        Args:
            rpc_url: Web3 RPC endpoint URL
            routers: List of known DEX router addresses to monitor
            window_size: Number of blocks to analyze
            weights: Tuple of (mev_ratio_weight, sandwich_weight, clustering_weight)
        """
        self.w3 = Web3(Web3.HTTPProvider(rpc_url))
        self.routers = [r.lower() for r in routers]
        self.window_size = window_size
        self.weights = weights
        self.high_gas_threshold = 5.0  # Dynamic threshold for high-gas transactions
        
        if not self.w3.is_connected():
            raise ConnectionError(f"Failed to connect to RPC endpoint: {rpc_url}")
    
    def get_searcher_density(self) -> float:
        """
        Calculate real-time searcher density score
        
        Returns:
            Searcher density score (0.0 = no MEV activity, 1.0 = extreme MEV activity)
        """
        try:
            mev_ratio = self._calculate_mev_transaction_ratio()
            sandwich_score = self._calculate_sandwich_attack_score()
            clustering_score = self._analyze_bot_clustering()
            
            density_score = (
                self.weights[0] * mev_ratio +
                self.weights[1] * sandwich_score +
                self.weights[2] * clustering_score
            )
            
            return min(density_score, 1.0)
            
        except Exception as e:
            print(f"Error in get_searcher_density: {e}")
            return 0.5
    
    def _calculate_mev_transaction_ratio(self) -> float:
        """Calculate ratio of MEV transactions to total transactions"""
        try:
            blocks = self._fetch_recent_blocks()
            total_txs = 0
            mev_txs = 0
            
            for block in blocks:
                if not hasattr(block, 'transactions') or not block.transactions:
                    continue
                    
                total_txs += len(block.transactions)
                
                # Count transactions interacting with DEX routers as potential MEV
                for tx_hash in block.transactions:
                    tx = self.w3.eth.get_transaction(tx_hash)
                    if tx and tx.to and tx.to.lower() in self.routers:
                        mev_txs += 1
            
            if total_txs == 0:
                return 0.0
            
            return min(mev_txs / total_txs, 1.0)
            
        except Exception as e:
            print(f"Error in _calculate_mev_transaction_ratio: {e}")
            return 0.0
    
    def _calculate_sandwich_attack_score(self) -> float:
        """
        Detect sandwich attack likelihood via gas price clustering
        High variance in gas prices indicates potential sandwich attacks
        """
        try:
            blocks = self._fetch_recent_blocks()
            gas_prices = []
            
            for block in blocks:
                if not hasattr(block, 'transactions') or not block.transactions:
                    continue
                
                for tx_hash in block.transactions:
                    try:
                        tx = self.w3.eth.get_transaction(tx_hash)
                        if tx and tx.to and tx.to.lower() in self.routers:
                            gas_price_gwei = self.w3.from_wei(tx.gasPrice, 'gwei')
                            gas_prices.append(float(gas_price_gwei))
                    except Exception:
                        continue
            
            if len(gas_prices) < 2:
                return 0.0
            
            # Calculate coefficient of variation (CV)
            mean_gas = statistics.mean(gas_prices)
            std_gas = statistics.stdev(gas_prices)
            
            if mean_gas == 0:
                return 0.0
            
            cv = std_gas / mean_gas
            
            # Normalize CV to 0-1 range (CV > 1.0 indicates high variance)
            sandwich_score = min(cv, 1.0)
            
            return sandwich_score
            
        except Exception as e:
            print(f"Error in _calculate_sandwich_attack_score: {e}")
            return 0.0
    
    def _analyze_bot_clustering(self) -> float:
        """
        Analyze concentration of high-gas-paying addresses
        More concentrated activity indicates higher bot density
        """
        try:
            blocks = self._fetch_recent_blocks()
            high_gas_addresses: Set[str] = set()
            total_addresses: Set[str] = set()
            
            # Calculate average gas price first
            gas_prices = []
            for block in blocks:
                if not hasattr(block, 'transactions') or not block.transactions:
                    continue
                
                for tx_hash in block.transactions:
                    try:
                        tx = self.w3.eth.get_transaction(tx_hash)
                        if tx and tx.to and tx.to.lower() in self.routers:
                            gas_price_gwei = self.w3.from_wei(tx.gasPrice, 'gwei')
                            gas_prices.append(float(gas_price_gwei))
                    except Exception:
                        continue
            
            if not gas_prices:
                return 0.0
            
            avg_gas = statistics.mean(gas_prices)
            
            # Identify high-gas addresses
            for block in blocks:
                if not hasattr(block, 'transactions') or not block.transactions:
                    continue
                
                for tx_hash in block.transactions:
                    try:
                        tx = self.w3.eth.get_transaction(tx_hash)
                        if tx and tx.to and tx.to.lower() in self.routers:
                            total_addresses.add(tx['from'].lower())
                            
                            gas_price_gwei = self.w3.from_wei(tx.gasPrice, 'gwei')
                            if gas_price_gwei > avg_gas * self.high_gas_threshold:
                                high_gas_addresses.add(tx['from'].lower())
                    except Exception:
                        continue
            
            if len(total_addresses) == 0:
                return 0.0
            
            # Calculate clustering score
            clustering_ratio = len(high_gas_addresses) / len(total_addresses)
            
            # Normalize: cap at 50 active addresses for max score
            normalized_score = min(len(high_gas_addresses) / 50.0, 1.0)
            
            # Combine ratio and absolute count
            clustering_score = (clustering_ratio + normalized_score) / 2
            
            return min(clustering_score, 1.0)
            
        except Exception as e:
            print(f"Error in _analyze_bot_clustering: {e}")
            return 0.0
    
    def _fetch_recent_blocks(self) -> List:
        """Fetch recent blocks for analysis"""
        blocks = []
        current_block_number = self.w3.eth.block_number
        
        for i in range(self.window_size):
            if current_block_number - i >= 0:
                try:
                    block = self.w3.eth.get_block(current_block_number - i, full_transactions=False)
                    blocks.append(block)
                except Exception:
                    continue
            else:
                break
        
        return blocks
    
    def get_metrics(self) -> Dict:
        """Get detailed searcher density metrics for debugging/monitoring"""
        return {
            "searcher_density": self.get_searcher_density(),
            "mev_transaction_ratio": self._calculate_mev_transaction_ratio(),
            "sandwich_attack_score": self._calculate_sandwich_attack_score(),
            "bot_clustering_score": self._analyze_bot_clustering(),
            "window_size": self.window_size,
            "weights": self.weights,
            "monitored_routers": len(self.routers)
        }


if __name__ == '__main__':
    # Example usage
    routers = [
        '0x68b3465833fb72A70ecDF485E0e4C7bD8665Fc45',  # Uniswap V3 Router 2
        '0xE592427A0AEce92De3Edee1F18E0157C05861564',  # Uniswap V3 Router
        '0x1b02dA8Cb0d097eB8D57A175b88c7D8b47997506'   # SushiSwap Router
    ]
    
    # This would require a valid RPC URL
    # sensor = SearcherDensitySensor('https://arb1.arbitrum.io/rpc', routers)
    # print(sensor.get_metrics())

File: src/mev/sensors/SearcherDensitySensor.ts
/**
 * SearcherDensitySensor - Quantify MEV bot activity density
 * Ported from AxionCitadel's mev_risk_arb/sensors/searcher_density.py
 *
 * Calculates a 0-1 density score based on:
 * 1. MEV transaction ratio
 * 2. Sandwich attack indicators (gas price variance)
 * 3. Bot clustering (unique high-gas addresses)
 */

import { Provider, getAddress, TransactionResponse } from 'ethers';
// ethers namespace reserved for utilities
import type { ethers as _ethers } from 'ethers';
import { SensorReading } from '../types/TransactionType';

export interface DensityWeights {
  mevRatio: number;
  sandwich: number;
  clustering: number;
}

const DEFAULT_WEIGHTS: DensityWeights = {
  mevRatio: 0.5,
  sandwich: 0.3,
  clustering: 0.2,
};

// MEV-sensitive contracts (should be configurable)
const DEFAULT_MEV_CONTRACTS = {
  UNISWAP_V3_ROUTER: '0x68b3465833fb72A70ecfFf80892690ab4c15dEc1',
  SUSHI_ROUTER: '0x1b02dA8Cb0d097eB8D57A175b88c7D8b47997506',
  // Add more as needed
};

export class SearcherDensitySensor {
  private provider: Provider;
  private weights: DensityWeights;
  private lookbackBlocks: number;
  private mevContracts: Set<string>;

  constructor(
    provider: Provider,
    lookbackBlocks: number = 20,
    weights?: Partial<DensityWeights>,
    mevContracts?: string[]
  ) {
    this.provider = provider;
    this.lookbackBlocks = lookbackBlocks;
    this.weights = { ...DEFAULT_WEIGHTS, ...weights };

    // Initialize MEV contract addresses (checksummed)
    const contracts = mevContracts || Object.values(DEFAULT_MEV_CONTRACTS);
    this.mevContracts = new Set();

    // Only add valid checksummed addresses
    for (const addr of contracts) {
      try {
        this.mevContracts.add(getAddress(addr));
      } catch (_error) {
        console.warn(`Invalid MEV contract address ignored: ${addr}`);
      }
    }
  }

  /**
   * Get real-time searcher density score (0-1 scale)
   */
  async getDensityScore(): Promise<number> {
    try {
      const currentBlockNumber = await this.provider.getBlockNumber();

      // 1. MEV transaction ratio
      const mevRatio = await this.calculateMEVRatio(currentBlockNumber);

      // 2. Sandwich attack score proxy (gas price variance)
      const sandwichScore = await this.calculateSandwichScore(currentBlockNumber);

      // 3. Bot clustering
      const clustering = await this.calculateClustering(currentBlockNumber);

      // Weighted composition
      const densityScore =
        this.weights.mevRatio * mevRatio +
        this.weights.sandwich * sandwichScore +
        this.weights.clustering * clustering;

      return densityScore;
    } catch (error) {
      console.error('Error in getDensityScore:', error);
      // Fallback to neutral score
      return 0.5;
    }
  }

  /**
   * Get sensor reading with metadata
   */
  async getReading(): Promise<SensorReading> {
    const value = await this.getDensityScore();
    return {
      value,
      timestamp: Date.now(),
      source: 'SearcherDensitySensor',
    };
  }

  /**
   * Calculate MEV transaction ratio
   */
  private async calculateMEVRatio(currentBlockNumber: number): Promise<number> {
    try {
      let mevTxCount = 0;
      let totalTxCount = 0;

      for (let i = 0; i < this.lookbackBlocks; i++) {
        const blockNumber = currentBlockNumber - i;
        if (blockNumber < 0) break;

        const block = await this.provider.getBlock(blockNumber, true);
        if (!block) continue;

        totalTxCount += block.transactions.length;

        for (const txOrHash of block.transactions) {
          // In ethers v6, transactions can be TransactionResponse or string (hash)
          const tx =
            typeof txOrHash === 'string'
              ? await this.provider.getTransaction(txOrHash)
              : (txOrHash as TransactionResponse);
          if (!tx) continue;

          if (tx.to) {
            try {
              const checksummedTo = getAddress(tx.to);
              if (this.mevContracts.has(checksummedTo)) {
                mevTxCount++;
              }
            } catch {
              // Skip malformed addresses
              continue;
            }
          }
        }
      }

      return mevTxCount / Math.max(totalTxCount, 1);
    } catch (error) {
      console.warn('Error calculating MEV ratio:', error);
      return 0;
    }
  }

  /**
   * Calculate sandwich attack score (gas price variance as proxy)
   */
  private async calculateSandwichScore(currentBlockNumber: number): Promise<number> {
    try {
      const gasPrices: number[] = [];

      // Look at last 3 blocks for gas price variance
      const blocksToAnalyze = Math.min(this.lookbackBlocks, 3);
      for (let i = 0; i < blocksToAnalyze; i++) {
        const blockNumber = currentBlockNumber - i;
        if (blockNumber < 0) break;

        const block = await this.provider.getBlock(blockNumber, true);
        if (!block) continue;

        for (const txOrHash of block.transactions) {
          // In ethers v6, transactions can be TransactionResponse or string (hash)
          const tx =
            typeof txOrHash === 'string'
              ? await this.provider.getTransaction(txOrHash)
              : (txOrHash as TransactionResponse);
          if (!tx) continue;

          if (tx.gasPrice) {
            gasPrices.push(Number(tx.gasPrice));
          }
        }
      }

      if (gasPrices.length < 10) {
        return 0;
      }

      // Calculate coefficient of variation (stdev/mean)
      const mean = gasPrices.reduce((a, b) => a + b, 0) / gasPrices.length;
      if (mean === 0) {
        return 0;
      }

      const squaredDiffs = gasPrices.map((p) => Math.pow(p - mean, 2));
      const variance = squaredDiffs.reduce((a, b) => a + b, 0) / gasPrices.length;
      const stdev = Math.sqrt(variance);

      const coefficientOfVariation = stdev / mean;
      return Math.min(coefficientOfVariation, 1.0);
    } catch (error) {
      console.warn('Error calculating sandwich score:', error);
      return 0;
    }
  }

  /**
   * Calculate bot clustering (unique high-gas addresses)
   */
  private async calculateClustering(currentBlockNumber: number): Promise<number> {
    try {
      // First, get average gas price from latest block
      const latestBlock = await this.provider.getBlock('latest', true);
      if (!latestBlock || latestBlock.transactions.length === 0) {
        return 0;
      }

      const avgGasPrice =
        latestBlock.transactions
          .map((txOrHash) =>
            typeof txOrHash === 'string' ? null : (txOrHash as TransactionResponse)
          )
          .filter((tx): tx is TransactionResponse => tx !== null && tx.gasPrice !== null)
          .reduce((sum, tx) => sum + Number(tx.gasPrice), 0) /
        Math.max(
          latestBlock.transactions.filter(
            (txOrHash) => typeof txOrHash !== 'string' && (txOrHash as TransactionResponse).gasPrice
          ).length,
          1
        );

      const highGasThreshold = avgGasPrice * 5; // 5x average
      const botAddresses = new Set<string>();

      // Analyze last 3 blocks for high-gas transactions
      const blocksToAnalyze = Math.min(this.lookbackBlocks, 3);
      for (let i = 0; i < blocksToAnalyze; i++) {
        const blockNumber = currentBlockNumber - i;
        if (blockNumber < 0) break;

        const block = await this.provider.getBlock(blockNumber, true);
        if (!block) continue;

        for (const txOrHash of block.transactions) {
          // In ethers v6, transactions can be TransactionResponse or string (hash)
          const tx =
            typeof txOrHash === 'string'
              ? await this.provider.getTransaction(txOrHash)
              : (txOrHash as TransactionResponse);
          if (!tx) continue;

          if (tx.from && tx.gasPrice && Number(tx.gasPrice) > highGasThreshold) {
            try {
              const checksummedFrom = getAddress(tx.from);
              botAddresses.add(checksummedFrom);
            } catch {
              // Skip malformed addresses
              continue;
            }
          }
        }
      }

      // Normalize based on expected max number of bot addresses (50 is arbitrary)
      const clustering = Math.min(botAddresses.size / 50.0, 1.0);
      return clustering;
    } catch (error) {
      console.warn('Error calculating clustering:', error);
      return 0;
    }
  }

  /**
   * Add MEV contract address to monitor
   */
  addMEVContract(address: string): void {
    this.mevContracts.add(getAddress(address));
  }

  /**
   * Remove MEV contract address from monitoring
   */
  removeMEVContract(address: string): void {
    this.mevContracts.delete(getAddress(address));
  }
}
